### DISCUSSION TITLE
Reset traverse scope id

### DISCUSSION THREAD

**Discussion by PhilippeProulx** [babel/babel#13944] [Q&A] [answered]
Is there a way to reset the internal scope id counter when using traverse on unrelated ast?

```
const ast = parse("let a;");
const ast2 = parse("let b;");

traverse(ast, {VariableDeclarator: (path) => console.log(path.scope.uid)}); // 0
traverse(ast2, {VariableDeclarator: (path) => console.log(path.scope.uid)});  // 1 (would like to have 0 here)

**Comment by JLHwung** [comment, marked as answer]
Currently no, unless you run Babel in different processes. The scope uid base counter is defined in top level

https://github.com/babel/babel/blob/74bec41db58d8911ac61197e7818970bd4880822/packages/babel-traverse/src/scope/index.ts#L362

and will be incremented whenever a new scope is created.

What is your use case on resetting the counter?

**Reply by PhilippeProulx** [reply]
Thank you for the anwser!

My use case is probably unusual in that babel *parse/traverse/generate* is used in production to transform some user-provided code where the user then selects specific variables from his code to activate debugger-like functionnalities on them.

This "variable selection" from the user-provided code would be done by combining the *name* of the variable and its *scope id* in order to support cases where multiple variables in different scope share the same name.

This means the *scope id* would need to be stable accross multiple *traverse* calls since the user can edit the code, thus trigger *parse/traverse/generate* calls more than once in a given session.

The *line number* or *start position* of the *variable declaration* node could be used instead, but would be more prone to changes and issues than *scope id* when the user edits the code. I am also considering using their rank (1st, 2nd...) instead of *scope id* but this option would also have its own issues.

In case it can help someone else with a similar problem, my current solution is to save the first *scope id* of a given *traverse* call and keep it as a *scope id offset* to be used later to calculate the "relative/stable" *scope id* that I need from the cumulative one given by *traverse*.
